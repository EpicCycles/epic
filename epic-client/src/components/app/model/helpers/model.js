import {
    findObjectWithKey,
    generateRandomCode,
    isItAnObject,
    removeKey,
    updateObject,
    updateObjectInArray
} from "../../../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../../../helpers/constants";

export const eliminateReadOnlyFields = (fieldList) => {
    return fieldList.filter(field => (! field.readOnly));
};

export const checkForChanges = (fieldList, existingObject, newValues) => {
    return eliminateReadOnlyFields(fieldList).some(field => {
        return existingObject[field.fieldName] !== newValues[field.fieldName];
    })
};

export const addFieldToState = (initialState, fieldList, fieldName, input) => {
    let finalState = initialState;
    fieldList.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            finalState = applyFieldValueToModel(initialState, field, input);
            return true;
        }
        return false;
    });
    return finalState;
};


export const getModelKey = modelInstance => {
    if (!modelInstance) return NEW_ELEMENT_ID;
    if (modelInstance.id) return modelInstance.id;
    if (modelInstance.dummyKey) return modelInstance.dummyKey;
    return NEW_ELEMENT_ID;
};

export const isModelValid = (modelInstance) => {
    return !(modelInstance.error || isItAnObject(modelInstance.error_detail));
};

export const applyFieldValueToModel = (modelInstance, field, value) => {
    let updatedModelInstance = updateObject(modelInstance);
    updatedModelInstance[field.fieldName] = value;
    updatedModelInstance.changed = true;
    if (!updatedModelInstance.error_detail) updatedModelInstance.error_detail = {};
    if (field.required && !value) {
        updatedModelInstance.error_detail[field.fieldName] = field.error;
        return updatedModelInstance;
    } else if (field.validator) {
        const error = field.validator(value, modelInstance);
        if (error) {
            updatedModelInstance.error_detail[field.fieldName] = error;
            return updatedModelInstance;
        }
    }
    updatedModelInstance.error_detail = removeKey(updatedModelInstance.error_detail, field.fieldName);
    return updatedModelInstance;
};

export const getAttribute = (modelFields, fieldName) => {
    let attribute;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            attribute = field.fieldName;
            return true;
        }
        return false;
    });
    return attribute;
};

export const getField = (modelFields, fieldName) => {
    let modelField;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            modelField = field;
            return true;
        }
        return false;
    });
    return modelField;
};

export const updateModelChanges = (modelInstance, attribute, fieldValue) => {
    let changes = updateObject(modelInstance.changes);
    if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
        changes[attribute] = fieldValue;
    } else {
        changes = removeKey(changes, attribute);
    }
    return updateObject(modelInstance, { changes })
};
export const updateModelWithChanges = (modelInstance, modelFields, fieldName, fieldValue) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        return updateModelChanges(modelInstance, attribute, fieldValue)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelInstance;
};
export const updateModelArrayWithChanges = (modelArray, modelFields, fieldName, fieldValue, componentKey) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        let modelInstance = findObjectWithKey(modelArray, componentKey);
        let currentChanges = modelInstance.changes || {};
        if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
            currentChanges[attribute] = fieldValue;
        } else {
            currentChanges = removeKey(currentChanges, attribute);
        }
        modelInstance.changes = currentChanges;
        return updateObjectInArray(modelArray, modelInstance, componentKey)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelArray;
};
export const updateModel = (model, modelFields, fieldName, fieldValue, componentKey) => {
    const modelField = getField(modelFields, fieldName);
    if (modelField) return applyFieldValueToModel(model, modelField, fieldValue);
    return model;
};

export const displayModelErrorSummary = (model, modelFields) => {
    let displayErrors = model.error;
    const fieldErrors = isItAnObject(model.error_detail) ? model.error_detail : {};
    for (var property in fieldErrors) {
        const field = getField(modelFields, property);
        if (!displayErrors) {
            displayErrors = "";
        } else {
            displayErrors += '<br>';
        }
        displayErrors += field.header + ": " + fieldErrors[property].join(" ");
    }

    return displayErrors;
};

export const getNameForValue = (value, sourceArray) => {
    if (!value) return undefined;
    let name = value;

    sourceArray.some(pair => {
        if (pair.value === value) {
            name = pair.name;
            return true;
        }
        return false;
    });
    return name;
};

export const hasErrors = (model) => {
    if (model.error && model.error.length > 0) return true;
    if (isItAnObject(model.error_detail)) return true;
    return false;
};

export const addErrorDetail = (error_detail = {}, field, error) => {
      let updated_error_detail = updateObject(error_detail);
  if (!Array.isArray(updated_error_detail[field])) {
         updated_error_detail[field] = [];
  }
  updated_error_detail[field].push(error);
  return updated_error_detail;
};

export const createNewModelInstance = () => {
    return {dummyKey: generateRandomCode()};
};
export const matchesModel = (persistedModel, modelFields, modelToCheck) => {
     return eliminateReadOnlyFields(modelFields).every(field => {
        return persistedModel[field.fieldName] === modelToCheck[field.fieldName];
    })
};