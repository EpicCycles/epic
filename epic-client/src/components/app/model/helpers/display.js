export const fixedHeaderClassname = (lockColumn) => {
    if (lockColumn) return "grid-header--fixed-left";
    return "";
};
export const fixedDetailsClassname = (lockColumn) => {
    if (lockColumn) return "grid-item--fixed-left";
    return "";
};
export const gridHeaderClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item ${fixedDetailsClassname(shouldLock)}`;
};