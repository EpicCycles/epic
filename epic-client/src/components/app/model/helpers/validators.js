import {INVALID_POSTCODE} from "./error";
import {POSTCODE_RULES} from "./constants";
export const validateData = (fieldList, currentValues) => {
    let errors = {};
    fieldList.forEach(field => {
        if (field.required) {
            if (!currentValues[field.fieldName]) {
                errors[field.fieldName] = field.error;
            }
        }
    });
    return errors;
};

export const validateEmailFormat = (email) => {
    const emailPattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return emailPattern.test(email);
};

export const validatePostcodeAndReturnError = (postcode, model = {}) => {
    let error;
    const postcodeRule = POSTCODE_RULES.filter(rule => rule.countryCode === model.country);
    if (postcodeRule[0]) {
        const OK = RegExp(postcodeRule[0].regex).exec(postcode);
        if (!OK) error = INVALID_POSTCODE + postcodeRule[0].display;
    }
    return error;
};