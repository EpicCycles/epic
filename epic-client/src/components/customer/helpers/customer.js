import {findIndexOfObjectWithId, generateRandomCode, updateObject} from "../../../helpers/utils";

export const updateCustomerBasic = (customer = {}, updatedValues) => {
    return updateObject(customer, updatedValues);
};

export const updateCustomerPhoneList = (customer = {}, phones) => {
    let updatedCustomer = updateObject(customer);
    updatedCustomer.phones = phones;
    if (customer.phones.length < phones.length) updatedCustomer.newPhone = {"dummyKey": generateRandomCode()};
    return updatedCustomer;
};

export const updateCustomerAddressList = (customer = {}, addresses) => {
    let updatedCustomer = updateObject(customer);
    updatedCustomer.addresses = addresses;
    if (customer.addresses && (customer.addresses.length < addresses.length)) updatedCustomer.newAddress = {"dummyKey": generateRandomCode()};
    return updatedCustomer;
};

export const customerAddErrorForAddress = (customer = {}, customerAddress) => {

    let updatedCustomer = updateObject(customer);

    const addressId = customerAddress.id;
    if (addressId && updatedCustomer.addresses) {
        const indexOfAddress = findIndexOfObjectWithId(updatedCustomer.addresses, addressId);
        updatedCustomer.addresses[indexOfAddress].error = true;
    } else {
        updatedCustomer.newAddress = customerAddress;
        updatedCustomer.newAddress.error = true;
    }
    return updatedCustomer;
};

export const customerAddErrorForPhone = (customer = {}, customerPhone) => {
    let updatedCustomer = updateObject(customer);

    const phoneId = customerPhone.id;
    if (phoneId && updatedCustomer.phones) {
        const indexOfPhone = findIndexOfObjectWithId(updatedCustomer.phones, phoneId);
        updatedCustomer.phones[indexOfPhone].error = true;
    } else {
        updatedCustomer.newPhone = customerPhone;
        updatedCustomer.newPhone.error = true;
    }
    return updatedCustomer;
};

