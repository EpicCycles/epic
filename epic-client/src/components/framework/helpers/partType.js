import {updateObject} from "../../../helpers/utils";
import {PART_TYPE_NAME_MISSING} from "../../app/model/helpers/error";
import {NEW_ELEMENT_ID} from "../../../helpers/constants";

export const getPartTypeName = (partTypeId, sections) => {
    if (!partTypeId) return undefined;
    const partType = getPartType(partTypeId, sections);
    if (partType) return partType.name;
    return 'Unknown Part Type';
};
export const getPartType = (partTypeId, sections) => {
    if (!partTypeId) return undefined;
    let partType;
    sections.some(section => {
        return section.partTypes.some(aPartType => {
            if (aPartType.id === partTypeId) {
                partType = aPartType;
                return true;
            }
            return false;
        });
    });
    return partType;
};
export const getPartTypeAttributeFields = (partTypeId, sections) => {
    const partType = getPartType(partTypeId, sections);
    if (partType) return partType.attributes;
    return [];
};
export const processPartTypeValueChanges = (partType, componentKey, fieldName, input) => {
    const updatedPartType = updateObject(partType);
    if (fieldName.startsWith('name')) updatedPartType.name = input;
    if (!updatedPartType.name) {
        updatedPartType.error = true;
        updatedPartType.error_detail = PART_TYPE_NAME_MISSING;
    } else {
        updatedPartType.error = false;
        updatedPartType.error_detail = "";
    }
    if (fieldName.startsWith('description')) updatedPartType.description = input;
    if (fieldName.startsWith('can_be_substituted')) updatedPartType.can_be_substituted = input;
    if (fieldName.startsWith('can_be_omitted')) updatedPartType.can_be_omitted = input;
    if (fieldName.startsWith('customer_visible')) updatedPartType.customer_visible = input;
    if (fieldName.startsWith('attributes')) updatedPartType.attributes = input;
    if (fieldName.startsWith('synonyms')) updatedPartType.synonyms = input;
    if (fieldName.startsWith('detail')) updatedPartType._detail = input;
    if (componentKey === NEW_ELEMENT_ID) updatedPartType.dummyKey = NEW_ELEMENT_ID;

    updatedPartType.changed = true;
    return updatedPartType;
};
export const doesFieldMatchPartType = (partType, fieldName) => {
    const fieldNameLower = fieldName.toLowerCase();
    if (partType.name.toLowerCase() === fieldNameLower) {
        return true;
    } else {
        return partType.synonyms.some(synonym => {
            return (synonym.name.toLowerCase() === fieldNameLower);
        });
    }
};