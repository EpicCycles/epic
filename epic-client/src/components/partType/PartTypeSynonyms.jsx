import React from "react";
import FormTextInput from "../../common/FormTextInput";
import {findIndexOfObjectWithKey, generateRandomCode} from "../../helpers/utils";
import {Icon} from "semantic-ui-react";
import {NEW_ELEMENT_ID} from "../../helpers/constants";

class PartTypeSynonyms extends React.Component {
    handleInputChange = (fieldName, input) => {
        const fields = fieldName.split('_');
        const synonymKey = fields[1];
        const synonymsWithUpdates = this.props.synonyms.slice();
        const synonymToUpdateIndex = findIndexOfObjectWithKey(synonymsWithUpdates, synonymKey);

        if (synonymToUpdateIndex > -1) {
            if (input) {
                synonymsWithUpdates[synonymToUpdateIndex].shortName = input;
                synonymsWithUpdates[synonymToUpdateIndex].error = false;
                synonymsWithUpdates[synonymToUpdateIndex].error_detail = "";
            } else {
                synonymsWithUpdates[synonymToUpdateIndex].error = true;
                synonymsWithUpdates[synonymToUpdateIndex].error_detail = "A value is required for the synonym";
            }
            synonymsWithUpdates[synonymToUpdateIndex].changed = true;
        } else if (input) {
            synonymsWithUpdates.push({
                "dummyKey": NEW_ELEMENT_ID,
                "shortName": input
            });
        }

        this.props.handlePartTypeChange(`synonyms_${this.props.partTypeKey}`, synonymsWithUpdates);
    };
    addAnother = () => {
        const synonymsWithUpdates = this.props.synonyms.slice();
        const synonymToUpdateIndex = findIndexOfObjectWithKey(synonymsWithUpdates, NEW_ELEMENT_ID);
        synonymsWithUpdates[synonymToUpdateIndex].dummyKey = generateRandomCode();
        this.props.handlePartTypeChange(`synonyms_${this.props.partTypeKey}`, synonymsWithUpdates);
    };

    handleInputClear = (fieldName) => {
        const fields = fieldName.split('_');
        const synonymKey = fields[1];
        const synonymsWithUpdates = this.props.synonyms.slice();
        if (synonymKey !== "new") {
            const synonymToUpdateIndex = findIndexOfObjectWithKey(synonymsWithUpdates, synonymKey);
            if (synonymToUpdateIndex > -1) {
                if (window.confirm("Please confirm that you want to delete this Attribute")) {
                    synonymsWithUpdates[synonymToUpdateIndex].delete = true;
                    this.props.handlePartTypeChange(`synonyms_${this.props.partTypeKey}`, synonymsWithUpdates);
                }
            }
        }
    };

    render() {
        const { partTypeKey, synonyms } = this.props;
        const synonymsToUse = synonyms ? synonyms.filter(synonym => !(synonym.delete || (synonym.dummyKey === NEW_ELEMENT_ID))) : [];
        const newSynonym = synonyms ? synonyms.filter(synonym => (synonym.dummyKey === NEW_ELEMENT_ID)) : [];
        let newSynonymDisplay = (newSynonym.length > 0) ? newSynonym[0] : {};
        return <table>
            <tbody>
            {synonymsToUse.map((synonym) => {
                const componentKey = synonym.id ? synonym.id : synonym.dummyKey;
                const className = synonym.error ? "error" : "";
                const rowTitle = synonym.error ? synonym.error_detail : "";
                return <tr
                    key={`synonym_${componentKey}`}
                    className={className}
                    title={rowTitle}
                >
                    <td><FormTextInput
                        placeholder="add new"
                        fieldName={`synonymValue_${componentKey}`}
                        value={synonym.shortName}
                        onChange={this.handleInputChange}
                        onClick={this.handleInputClear}
                    /></td>
                </tr>
            })}

            <tr key={`newSynonym${partTypeKey}`}>
                <td><FormTextInput
                    placeholder="add new"
                    fieldName="synonymValue_new"
                    value={newSynonymDisplay.shortName ? newSynonymDisplay.shortName : ""}
                    onChange={this.handleInputChange}
                    onClick={this.handleInputClear}
                /></td>
                <td>
                    <Icon
                        name="add"
                        onClick={this.addAnother}
                        title="confirm new Synonym"
                    />
                </td>
            </tr>
            </tbody>
        </table>;
    }
}

export default PartTypeSynonyms;