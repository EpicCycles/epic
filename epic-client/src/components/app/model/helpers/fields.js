import {
    ADDRESS_MISSING,
    BRAND_MISSING, BRAND_NAME_MISSING,
    BUNDLE_NAME_MISSING, COUNTRY_MISSING,
    FRAME_NAME_MISSING,
    MODEL_NAME_MISSING,
    PART_MISSING,
    PART_NAME_MISSING,
    PART_TYPE_MISSING,
    PRODUCTS_MISSING,
    SUPPLIER_MISSING
} from "./error";
import {validatePostcodeAndReturnError} from "./validators";

// TODO customer fields including email with email validator
export const ADDRESS1 = "address1";
export const ADDRESS2 = "address2";
export const ADDRESS3 = "address3";
export const ADDRESS4 = "address4";
export const COUNTRY = "country";
export const POSTCODE = "postcode";
export const BRAND = "brand";
export const SUPPLIER = "supplier";
export const PART = "part";
export const LINK = "link";
export const BUNDLE = "bundle_name";
export const PRODUCT_CODE = "product_code";
export const PART_TYPE = "partType";
export const BRAND_NAME = "brand_name";
export const BIKE_BRAND = "bike_brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const PART_NAME = "part_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "rrp";
export const EPIC_PRICE = "epic_price";
export const TICKET_PRICE = "ticket_price";
export const TRADE_PRICE = "trade_price";
export const CLUB_PRICE = "club_price";
export const FITTED_PRICE = "fitted_price";
export const TRADE_IN_PRICE = "trade_in_value";
export const SIZES = "sizes";
export const STANDARD = "standard";
export const STOCKED = "stocked";
export const PRODUCTS = "products";
export const TEXT = "text";
export const NUMBER = "number";
export const CURRENCY = "currency";
export const TEXT_AREA = "textArea";
export const CHECKBOX = "checkbox";

// TODO add COUNTRY type and fields for address
// TODO consider having a function passed for validator
export const ADDRESS1_FIELD = {
    fieldName: ADDRESS1,
    type: TEXT,
    length: 100,
    header: "Address",
    required: true,
    error: ADDRESS_MISSING
};
export const ADDRESS2_FIELD = {
    fieldName: ADDRESS2,
    type: TEXT,
    length: 100,
    header: "Line 2",
};
export const ADDRESS3_FIELD = {
    fieldName: ADDRESS3,
    type: TEXT,
    length: 100,
    header: "Line 3",
};
export const ADDRESS4_FIELD = {
    fieldName: ADDRESS4,
    type: TEXT,
    length: 100,
    header: "Line 4",
};
export const COUNTRY_FIELD = {
    fieldName: COUNTRY,
    header: "Country",
    type: COUNTRY,
    required: true,
    error: COUNTRY_MISSING
};
export const POSTCODE_FIELD = {
    fieldName: POSTCODE,
    type: TEXT,
    length: 20,
    header: "Postcode",
    validator: validatePostcodeAndReturnError,
    validatorAdditionalFields: [COUNTRY]
};
export const BRAND_FIELD = {
    fieldName: BRAND,
    header: "Brand",
    synonyms: [BRAND, 'manufacturer'],
    required: true,
    error: BRAND_MISSING,
    type: BRAND
};
export const PRODUCTS_FIELD = {
    fieldName: PRODUCTS,
    header: "Products",
    synonyms: [],
    required: true,
    error: PRODUCTS_MISSING,
    type: PRODUCTS
};
export const SUPPLIER_FIELD = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    required: true,
    error: SUPPLIER_MISSING,
    type: SUPPLIER
};
export const SUPPLIER_FIELD_OPTIONAL = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    type: SUPPLIER
};
export const PART_FIELD = {
    fieldName: PART,
    header: "Part",
    synonyms: [PART],
    required: true,
    error: PART_MISSING,
    type: PART
};
export const FRAME_NAME_FIELD = {
    fieldName: FRAME_NAME,
    header: "Frame Name",
    synonyms: [],
    required: true,
    error: FRAME_NAME_MISSING
};
export const LINK_FIELD = {
    fieldName: LINK,
    header: "URL",
    synonyms: [],
};
export const BRAND_NAME_FIELD = {
    fieldName: BRAND_NAME,
    header: "Brand Name",
    synonyms: [],
    required: true,
    error: BRAND_NAME_MISSING
};
export const BUNDLE_NAME_FIELD = {
    fieldName: BUNDLE,
    header: "Bundle Name",
    synonyms: [],
    required: true,
    error: BUNDLE_NAME_MISSING
};
export const MODEL_NAME_FIELD = {
    fieldName: MODEL_NAME,
    header: "Model Name",
    synonyms: [],
    required: true,
    error: MODEL_NAME_MISSING,
    type: TEXT,
    length: 100
};
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"],
    type: TEXT,
    length: 100
};
export const DESCRIPTION_FIELD = {
    fieldName: DESCRIPTION,
    header: "Description",
    synonyms: [DESCRIPTION, "desc"],
    type: TEXT_AREA,
    length: 400
};
export const PRODUCT_CODE_FIELD = {
    fieldName: PRODUCT_CODE,
    header: "Product Code",
    synonyms: [PRODUCT_CODE],
    type: TEXT,
    length: 30
};
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "RRP",
    synonyms: ["price", "selling price", "srp", "rrp", "sell price", "retail price"],
    type: CURRENCY,
    length: 10
};
export const EPIC_PRICE_FIELD = {
    fieldName: EPIC_PRICE,
    header: "Epic Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const CLUB_PRICE_FIELD = {
    fieldName: CLUB_PRICE,
    header: "Club Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};

export const FITTED_PRICE_FIELD = {
    fieldName: FITTED_PRICE,
    header: "Fitted Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TRADE_PRICE_FIELD = {
    fieldName: TRADE_PRICE,
    header: "Trade Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TICKET_PRICE_FIELD = {
    fieldName: TICKET_PRICE,
    header: "Ticket Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"],
    type: TEXT,
    length: 100
};
export const PART_TYPE_FIELD = {
    fieldName: PART_TYPE,
    header: "Part Type",
    required: true,
    error: PART_TYPE_MISSING,
    type: PART_TYPE,
};
export const PART_NAME_FIELD = {
    fieldName: PART_NAME,
    header: "Part Name",
    required: true,
    error: PART_NAME_MISSING
};
export const TRADE_IN_FIELD = {
    fieldName: TRADE_IN_PRICE,
    header: "Trade In Â£",
    type: CURRENCY,
    length: 10
};
export const STOCKED_FIELD = {
    fieldName: STOCKED,
    header: "Stocked",
    type: CHECKBOX
};
export const STANDARD_FIELD = {
    fieldName: STANDARD,
    header: "Standard",
    type: CHECKBOX
};
export const BIKE_BRAND_FIELD = {
    fieldName: BIKE_BRAND,
    header: "Bike Brand",
    type: CHECKBOX
};
export const customerAddressFields = [ADDRESS1_FIELD, ADDRESS2_FIELD, ADDRESS3_FIELD, ADDRESS4_FIELD, COUNTRY_FIELD, POSTCODE_FIELD];
export const frameFields = [BRAND, FRAME_NAME];
export const brandFields = [BRAND_NAME_FIELD, BIKE_BRAND_FIELD, LINK_FIELD, SUPPLIER_FIELD_OPTIONAL];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, EPIC_PRICE_FIELD, CLUB_PRICE_FIELD, SIZES_FIELD];
export const partFields = [
    PART_TYPE_FIELD,
    BRAND_FIELD,
    PART_NAME_FIELD,
    TRADE_IN_FIELD,
    STANDARD_FIELD,
    STOCKED_FIELD
];
export const supplierProductFields = [
    SUPPLIER_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
    SELL_PRICE_FIELD,
    TRADE_PRICE_FIELD,
    CLUB_PRICE_FIELD
];
export const bundleFields = [
    BUNDLE_NAME_FIELD,
    PRODUCTS_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
];

