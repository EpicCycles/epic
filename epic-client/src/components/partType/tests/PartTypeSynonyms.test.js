import React from 'react';
import {NEW_ELEMENT_ID} from "../../../helpers/constants";
import PartTypeSynonyms from "../PartTypeSynonyms";

test("it displays synonyms", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
    />);
    expect(component).toMatchSnapshot();
});
test("it adds a change to an existing element by id", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    const handlePartTypeChange = jest.fn();
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
        handlePartTypeChange={handlePartTypeChange}
    />);
    component.instance().handleInputChange("shortName_21","new name");
    const expectedSynonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "new name", changed:true, error: false, error_detail:""},
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    expect(handlePartTypeChange.mock.calls.length).toBe(1);
    expect(handlePartTypeChange.mock.calls[0][0]).toEqual("synonyms_45");
    expect(handlePartTypeChange.mock.calls[0][1]).toEqual(expectedSynonyms);
});
test("it adds an error to an existing element by id", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    const handlePartTypeChange = jest.fn();
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
        handlePartTypeChange={handlePartTypeChange}
    />);
    component.instance().handleInputChange("shortName_21","");
    const expectedSynonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "", changed:true, error: true, error_detail:"A value is required for the synonym"},
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    expect(handlePartTypeChange.mock.calls.length).toBe(1);
    expect(handlePartTypeChange.mock.calls[0][0]).toEqual("synonyms_45");
    expect(handlePartTypeChange.mock.calls[0][1]).toEqual(expectedSynonyms);
});
test("it adds a change to an existing element by key", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    const handlePartTypeChange = jest.fn();
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
        handlePartTypeChange={handlePartTypeChange}
    />);
    component.instance().handleInputChange("shortName_dummy2","new name");
    const expectedSynonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "new name", changed:true, error: false, error_detail:""},
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    expect(handlePartTypeChange.mock.calls.length).toBe(1);
    expect(handlePartTypeChange.mock.calls[0][0]).toEqual("synonyms_45");
    expect(handlePartTypeChange.mock.calls[0][1]).toEqual(expectedSynonyms);
});
test("it adds new element when data keyed", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
    ];
    const handlePartTypeChange = jest.fn();
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
        handlePartTypeChange={handlePartTypeChange}
    />);
    component.instance().handleInputChange("shortName_not_found","new one");
    const expectedSynonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    expect(handlePartTypeChange.mock.calls.length).toBe(1);
    expect(handlePartTypeChange.mock.calls[0][0]).toEqual("synonyms_45");
    expect(handlePartTypeChange.mock.calls[0][1]).toEqual(expectedSynonyms);
});
test("it converts a new element to a dummy key to add another", () => {
    const partTypeKey = 45;
    const synonyms = [
        { id: 1, shortName: "synonym 1" },
        { id: 21, shortName: "synonym 21" },
        { dummyKey: 'dummy2', shortName: "synonym 2" },
        { dummyKey: NEW_ELEMENT_ID, shortName: "new one" }
    ];
    const handlePartTypeChange = jest.fn();
    const component = shallow(<PartTypeSynonyms
        partTypeKey={partTypeKey}
        synonyms={synonyms}
        handlePartTypeChange={handlePartTypeChange}
    />);
    component.instance().addAnother();
    expect(handlePartTypeChange.mock.calls.length).toBe(1);
    expect(handlePartTypeChange.mock.calls[0][0]).toEqual("synonyms_45");
    expect(handlePartTypeChange.mock.calls[0][1].length).toEqual(synonyms.length);
    expect(handlePartTypeChange.mock.calls[0][1][3].dummyKey).not.toEqual(NEW_ELEMENT_ID);
    expect(handlePartTypeChange.mock.calls[0][1][3].shortName).toEqual(synonyms[3].shortName);

});
