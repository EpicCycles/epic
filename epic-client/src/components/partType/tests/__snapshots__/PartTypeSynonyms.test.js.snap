// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it displays synonyms 1`] = `
<React.Fragment>
  <FormTextInput
    className=""
    dataType="text"
    error=""
    fieldName="synonymValue_1"
    key="synonymValue_1"
    label=""
    onChange={[Function]}
    onClick={[Function]}
    placeholder="add new"
    size={30}
    title=""
    value="synonym 1"
  />
  <FormTextInput
    className=""
    dataType="text"
    error=""
    fieldName="synonymValue_dummy2"
    key="synonymValue_dummy2"
    label=""
    onChange={[Function]}
    onClick={[Function]}
    placeholder="add new"
    size={30}
    title=""
    value="synonym 2"
  />
  <div>
    <FormTextInput
      className=""
      dataType="text"
      error=""
      fieldName="synonymValue_new"
      key="synonymValue_new"
      label=""
      onChange={[Function]}
      onClick={[Function]}
      placeholder="add new"
      size={30}
      title=""
      value="new one"
    />
    <Icon
      as="i"
      key="synonym_add"
      name="add"
      onClick={[Function]}
      title="confirm new Synonym"
    />
  </div>
</React.Fragment>
`;
